Add PGI compiler toolchain.

bartoldeman opened the pull request:
This adds a PGI compiler toolchain, see also pull request for https://github.com/hpcugent/easybuild-easyblocks/pull/658

boegel commented on easybuild/toolchains/compiler/pgi.py:
```diff
@@ -0,0 +1,71 @@
+##
+# Copyright 2015 Bart Oldeman
```
transfer copyright to Ghent University, or see https://github.com/hpcugent/easybuild-framework/blob/master/CONTRIBUTING.md

boegel replied:
nvm, triple-licensing also works

boegel commented on easybuild/toolchains/pgi.py:
```diff
@@ -0,0 +1,36 @@
+##
+# Copyright 2015 Bart Oldeman
```
transfer copyright to Ghent University, or see https://github.com/hpcugent/easybuild-framework/blob/master/CONTRIBUTING.md

boegel replied:
nvm, triple-licensed

boegel commented on easybuild/toolchains/compiler/pgi.py:
```diff
@@ -0,0 +1,71 @@
+
+
+class Pgi(Compiler):
+    """PGI compiler class
```
enhance?

boegel commented on easybuild/toolchains/compiler/pgi.py:
```diff
@@ -0,0 +1,71 @@
+    """PGI compiler class
+    """
+
+    COMPILER_MODULE_NAME = ['pgi']
```
PGI?

boegel commented on easybuild/toolchains/compiler/pgi.py:
```diff
@@ -0,0 +1,71 @@
+    COMPILER_UNIQUE_OPTION_MAP = {
+        'i8': 'i8',
+        'r8': 'r8',
+        'optarch': '', # default!
```
can you clarify this? the default mode of operation for PGI is to optimise for the system you're building on?

bartoldeman replied:
Yes, this is the default. I will put in a comment

boegel commented on easybuild/toolchains/compiler/pgi.py:
```diff
@@ -0,0 +1,71 @@
+        'precise': ['Mnoflushz'],
+        'defaultprec': ['Mflushz'],
+        'loose': ['Mfprelaxed'],
+        'veryloose': ['Mfprelaxed=div,order,intrinsic,recip,sqrt,rsqrt', 'Mfpapprox'],
```
any references for this?

bartoldeman replied:
e.g.:
http://www.pgroup.com/products/freepgi/freepgi_ref/ch02.html

boegel replied:
Please include these as comments above this line:

http://www.pgroup.com/products/freepgi/freepgi_ref/ch02.html#Mflushz
http://www.pgroup.com/products/freepgi/freepgi_ref/ch02.html#Mfprelaxed
http://www.pgroup.com/products/freepgi/freepgi_ref/ch02.html#Mfpapprox

boegel replied:
@bartoldeman please include the URLs mentioned as comments above this line?

bartoldeman replied:
done

boegel commented:
@georgets, @geimer, @rjeschmi: since you have looked into PGI support as well, please review? is a draft of what you were working on available somewhere?

boegel commented:
Jenkins: ok to test

rjeschmi commented:
I didn't get started on anything, but I ahve a need for this so will try to test.

rjeschmi commented:
I have tested this and didn't have any problems.

@bartoldeman might want to rebase to develop

boegel commented on easybuild/toolchains/compiler/pgi.py:
```diff
@@ -0,0 +1,75 @@
+    COMPILER_UNIQUE_OPTION_MAP = {
+        'i8': 'i8',
+        'r8': 'r8',
+        'optarch': '', # PGI by default generates code for the arch it is running on!
```
is there any way to disable this default behaviour, and specify the architecture to generate code for?

geimer replied:
The target architecture can be specified using the `-tp=<arch>` option.  PGI 14.1 supports the following architectures, newer versions may accept more (can't check atm):
```
    px              Generic x86 Processor
    k8              AMD64 Processor
    k8-64e          AMD64 Processor rev E or later, 64-bit mode
    barcelona       AMD Barcelona processor
    shanghai        AMD Shanghai processor
    istanbul        AMD Istanbul processor
    bulldozer       AMD Bulldozer processor
    piledriver      AMD Piledriver processor
    p7              Intel P7 Architecture (Pentium 4, Xeon, Centrino)
    core2           Intel Core-2 Architecture
    penryn          Intel Penryn Architecture
    nehalem         Intel Nehalem processor
    sandybridge     Intel SandyBridge processor
    haswell         Intel Haswell processor
    x64             Unified AMD/Intel mode; -tp=k8,p7
```

boegel replied:
@geimer: thanks for the feedback!

It would be interesting to somehow enforce picking one of these in case the `optarch` toolchain option is set to `False`.
The default is `True`, which (apparently) corresponds to not specifying any specific flag with PGI, but every now and then `False` is specified explicitly, so we should not just ignore that.

boegel replied:
Maybe we need to support a custom toolchain option for PGI-based toolchains, something like `target-platform`?

Or, we should reuse `optarch` for this? If `optarch` is `True`, we don't add any specific flags, if it's `False` we complain to specify the name of a target platform instead?

geimer replied:
`target-platform` would be interesting for other compilers as well, however, I thought the recommended way of handling this is the `--optarch=<arg>` command-line option.  Why not simply use a generic target like `px` or `x64` if `optarch` is `False`?  AFAIK, this is the default behavior of other compilers.  For PGI you have to explicitly request it.

boegel replied:
OK, that makes sense too, indeed. `x64` seems like a decent default indeed, and people can always override via `--optarch`.

@bartoldeman: are you up for tackling this yourself, i.e. using `-tp=x64` if `optarch` is set to `False`? See https://github.com/hpcugent/easybuild-framework/blob/master/easybuild/toolchains/compiler/craype.py#L100 for inspiration.

boegel replied:
Done below in `_set_compiler_flags`, thanks @bartoldeman.

ocaisa commented:
I spoke to someone from NVIDIA yesterday and he mentioned that C++ support in PGI depends on the underlying GCC. I think it might be wise to build up the PGI toolchain in the same way as the intel toolchain (bintuils -> GCC + bintuils -> PGI) to side step problems from the underlying OS (like RedHAt shipping an ancient GCC and a binutils that doesn't support modern architectures)

boegel commented:
@ocaisa: already done that way, see https://github.com/hpcugent/easybuild-easyconfigs/pull/1833

ocaisa commented:
eb --delete-lazy-comment

boegel commented:
@ocaisa: please make that a separate PR ;)

boegel commented:
Other stuff is coming in that is built on top of this (cfr. #1537), so we should really try and get this merged.

@bartoldeman: we haven't heard from you in a while on this, are you still willing to follow up on this, or should we take it from here?

bartoldeman commented:
Sorry for being silent on this. I am still willing to followup on this. I'll review and get back to you tomorrow.

boegel commented:
@bartoldeman Great, thanks for getting back to us on this!

bartoldeman commented:
I fixed the capitalizations and added `--optarch=GENERIC` support. I am a bit confused about how to make sure `optarch` is `False` works so that it then also adds -tp=x64, just like GENERIC.

bartoldeman commented:
I think it can work like this:
```python
    def _set_optimal_architecture(self):
        """Set -tp=x64 if optarch is False."""
        if not self.options.get('optarch', False):
            for optflag in self.COMPILER_OPT_FLAGS:
                self.COMPILER_UNIQUE_OPTION_MAP[optflag] += ' -tp=x64'
```

I will double check tomorrow.

bartoldeman commented:
`_set_optimal_architecture()` is called too early, before `self.options['optarch']==False` with the toolchainopts set that way.

I fixed that and verified it works by using `_set_compiler_flags()` instead.

jhein32 commented:
Hi,

I had a look at https://github.com/bartoldeman/easybuild-framework/blob/pgi/easybuild/toolchains/compiler/pgi.py (I hope the URL points to the right branch etc.).  I like to discuss/flag the following:

Looking into the PGI documentation http://www.pgroup.com/doc/pgiug.pdf (Table 1 in the preface) it is my understanding the PGI compiler should be invoked with pgf77 for F77 code and that pgfortran defaults to F2003.  They have an pgf90 command which I tend to use, that according to the man pages of PGI 15.10 is for f90 and f95 but they don't seem to push that in the PGI user guide any longer.  The current version of the compiler/pgi.py uses pgfortran for all Fortran compilers (F77, F90 and FC).  I am wondering whether  COMPILER_F77 = 'pgf77' is needed (line 79 of currrent version).

boegel commented:
@jhein32: the easiest way to make remarks is to do "inline comments", via https://github.com/hpcugent/easybuild-framework/pull/1342/files; hover of the line numbers, and you'll see a '+' appear; if you click it, you can add a line comment

jhein32 commented on easybuild/toolchains/compiler/pgi.py:
```diff
@@ -0,0 +1,92 @@
+    COMPILER_CC = 'pgcc'
+    COMPILER_CXX = 'pgc++'
+
+    COMPILER_F77 = 'pgfortran'
```
My understanding of the PGI userguide is that this should be 
```
Compiler_F77 = 'pgf77'
```
for Fortran 77.  See my entry in the comment section, but I am happy to be convinced differently.

boegel commented:
@bartoldeman Do you have time in the coming weeks to follow up on this and get this merged? I need to revisit the comments to see if there's actually anything left to do here, and I need to (finally) get a setup going so I can test PGI builds, but I'm really keen on getting this in ASAP.

Especially since @jgphpc is already building on top of it on Cray systems (cfr. #1537)...

boegel commented:
@damianam can you check whether this is exactly what you're using @ JSC? cc @ocaisa

bartoldeman commented:
@boegel Sorry for the delay. I will have time to follow up on this next week at the latest.

bartoldeman commented:
I rebased this PR and updated with the change from @jhein32 so pgf77 is used for F77.

damianam commented on easybuild/toolchains/pgi.py:
```diff
@@ -0,0 +1,39 @@
+@author: Bart Oldeman (McGill University, Calcul Quebec, Compute Canada)
+"""
+
+from easybuild.toolchains.compiler.pgi import Pgi
```
Shouldn't this import and use GCCcore as well? PGI compilers need an underlying GCC and I guess also binutils. Without GCCcore here any software installed there won't be available for PGI-based toolchains when using minimal toolchains.

bartoldeman replied:
yes, this means just catching up with what is happening in easybuild... I put this in the same way as for iccifort.py.

boegel replied:
Thanks @bartoldeman!

bartoldeman commented:
Here is the story with numactl (I mentioned that at the conf call):
any program that you compile with pgcc (with or without -mp) is linked to libnuma:
```
$ pgcc hello.c
$ ldd a.out 
        linux-vdso.so.1 =>  (0x00007ffc6676b000)
        libpgmp.so => /sb/software/CentOS-6/eb/software/Core/PGI/16.3-GCC-4.9.3-2.25/linux86-64/16.3/lib/libpgmp.so (0x00002ae2cc3ca000)
        libnuma.so.1 => /usr/lib64/libnuma.so.1 (0x0000003b96e00000)
        libpthread.so.0 => /lib64/libpthread.so.0 (0x0000003b95a00000)
        libpgc.so => /sb/software/CentOS-6/eb/software/Core/PGI/16.3-GCC-4.9.3-2.25/linux86-64/16.3/lib/libpgc.so (0x00002ae2cc66c000)
        libm.so.6 => /lib64/libm.so.6 (0x0000003b95e00000)
        libc.so.6 => /lib64/libc.so.6 (0x0000003b95600000)
        /lib64/ld-linux-x86-64.so.2 (0x0000003b95200000)
```
If the PGI installer detects /usr/lib64/libnuma.so.1 it will use that, if not, it will symlink linux86-64/16.3/lib/libnuma.so and ....so.1 to linux86-64/16.3/lib/libpgnuma.so. EB has its own numactl however. Perhaps we need numactl built with GCCcore?

valtandor commented:
How is this toolchain coming along? Does it need some additional testing?

boegel commented:
@valtandor things are slowly aligning to get this merged, see also https://github.com/hpcugent/easybuild/issues/211

A main issue on my end is that I need to set up the license that was provided to me by PGI, for testing purposes. That's less straightforward than I was hoping it would be, but I'm getting there...

boegel commented:
@bartoldeman do you mind updating this with latest `develop`, to trigger Travis testing your PR as well (we're working on moving from Jenkins to Travis, and I've enabled Travis tests as required to be able to merge PRs)?

Let me know if you need help with that, I can easily send you a PR for it and that's easier for you.

boegel commented:
All green, so this is good to go...

I'd like to get https://github.com/hpcugent/easybuild-easyconfigs/pull/2899 merge-ready first though.

It pretty much should be, but we've hit a snag with the module naming scheme @bartoldeman is using, causing the OpenMPI sanity check to fail. I'd like to figure that out first...

@ocaisa, @damianam: have you guys ever used PGI with `ToolchainMNS`, or another hierarchical module naming scheme?

boegel commented:
#1754 fixes the bug that was holding this PR back, and is merged now, so this is good to go as well

Thanks a lot for your work on this @bartoldeman, and for your patience in following this up; and of course also thanks to everyone else who providing feedback and testing!
